test-data/unit/check-overloading.test

```diff
@@ -1400,8 +1400,8 @@ a: Any
 b: List[Any]
 c: List[str]
 d: List[int]
-reveal_type(f(a)) # E: Revealed type is 'builtins.list[Any]'
-reveal_type(f(b))  # E: Revealed type is 'builtins.list[Any]'
```

Comment:
TBH, I don't like this change. I think the original issue only appears when one of the returns is (an explicit or inferred) `Any`. In this case `List[Any]` is perfectly safe. So a better fix would be to check if there are `Any`s among return and return `Any` in this case. Otherwise keep the current logic.

Reply:
I'm not so sure anymore if that'll necessarily be safe in all cases though. For example:

    from typing import Any, overload, TypeVar, Generic

    T = TypeVar('T')

    class Wrapper(Generic[T]): pass

    class A(Generic[T]):
        @overload
        def f(self, x: int) -> Wrapper[T]: ...
        @overload
        def f(self, x: slice) -> Wrapper[A[T]]: ...
        def f(self, x): ...

    i: Any
    a: A[Any]
    reveal_type(a.f(i))  # Wrapper[A[Any]], but should be Wrapper[Any] or Any

It's basically the same example from https://github.com/python/mypy/issues/5250, except I added a Wrapper class around the return types and we get the same sort of problem, except with no top-level Any.

Reply:
Addition: if you prefer it is totally OK to just use heuristics to find if there is the least precise type among all matching returns, if we can't find such, just return `Any`. I am fine if we will just cover the above case and the test case I mentioned. (I could imagine a careful treatment of all possibilities may be tedious).

Reply:
Wouldn't that algorithm cause a regression with https://github.com/python/mypy/issues/5232 though? (e.g. suppose we replace B with 'unicode' and C with 'bytes' in your test case)

Reply:
Sigh. Yes, this would be problematic. But at least can we "save" the test case? For example we allow a fallback option if "erased" types (with all args replace with `Any`) are the same for all matching returns, then we return the erased type?

Reply:
Hmm, I think that would probably work. (Or more precisely, I can't think of a way that would break yet, lol)

I'll implementing that change in a bit.